src/test/java/htsjdk/variant/bcf2/BCF2WriterUnitTest.java

```diff
@@ -0,0 +1,301 @@
+        }
+
+        VariantContextTestProvider.VariantContextContainer container;
+
+        try (final PositionalBufferedStream headerPbs = new PositionalBufferedStream(new FileInputStream(bcfOutputFile));
+        final PositionalBufferedStream bodyPbs = new PositionalBufferedStream(new FileInputStream(bcfOutputHeaderlessFile))) {
+
+            BCF2Codec codec = new BCF2Codec();
+            FeatureCodecHeader readHeader = codec.readHeader(headerPbs);
+            final VCFHeader vcfHeader = (VCFHeader) readHeader.getHeaderValue();
+            // we use the header information read from identical file with header+body to read just the body of second file
+            container =
+                    new VariantContextTestProvider.VariantContextContainer(vcfHeader, new VariantContextTestProvider.VCIterable(codec, vcfHeader) {
+                        @Override
+                        public boolean hasNext() {
+                            try {
+                                return !bodyPbs.isDone();
+                            } catch (IOException e) {
+                                throw new RuntimeException(e);
+                            }
+                        }
+
+                        @Override
+                        public Object nextSource() {
+                            return bodyPbs;
+                        }
+                    });
+
+            int counter = 0;
+            final Iterator<VariantContext> it = container.getVCs().iterator();
+            while (it.hasNext()) {
+                it.next();
+                counter++;
+            }
+            Assert.assertEquals(counter, 2);
+        }
+
+    }
+
+    @Test(expectedExceptions = IllegalStateException.class)
+    public void testWriteHeaderTwice() {
+        final File bcfOutputFile = VariantBaseTest.createTempFile("testWriteAndReadVCF.", ".bcf");
+
+        Tribble.indexFile(bcfOutputFile).deleteOnExit();
+        Set<VCFHeaderLine> metaData = new HashSet<VCFHeaderLine>();
+        Set<String> additionalColumns = new HashSet<String>();
+        final SAMSequenceDictionary sequenceDict = createArtificialSequenceDictionary();
+        final VCFHeader header = createFakeHeader(metaData, additionalColumns, sequenceDict);
+
+        // prevent writing header twice
+        final VariantContextWriter writer1 = new VariantContextWriterBuilder()
+                .setOutputFile(bcfOutputFile).setReferenceDictionary(sequenceDict)
+                .setOptions(EnumSet.of(Options.INDEX_ON_THE_FLY))
+                .build();
+        writer1.writeHeader(header);
+        writer1.writeHeader(header);
+        writer1.add(createVC(header));
+        writer1.close();
+    }
+
+    @Test(expectedExceptions = IllegalStateException.class)
+    public void testChangeHeaderAfterWritingHeader() {
+        final File bcfOutputFile = VariantBaseTest.createTempFile("testWriteAndReadVCF.", ".bcf");
+
+        Tribble.indexFile(bcfOutputFile).deleteOnExit();
```

Comment:
The reason I suggested creating a temporary directory in the earlier review pass was so this kind of call would be unnecessary.